src/app/Dashboard/Charts/ChartCard.tsx

```diff
@@ -0,0 +1,367 @@
+    let height: number;
+    switch (props.chartKind) {
+      case 'Core Count':
+        height = 250;
+        break;
+      default:
+        height = 380;
+        break;
+    }
+    return hasRecording
+      ? {
+          height,
+        }
+      : {};
+  }, [hasRecording, props.chartKind]);
+
+  const resyncButton = React.useMemo(() => {
+    return (
+      <Button
+        key={0}
+        aria-label={t('CHART_CARD.BUTTONS.SYNC.LABEL', { chartKind: props.chartKind })}
+        onClick={updateKey}
+        variant="plain"
+        icon={<SyncAltIcon />}
+        isDisabled={!chartSrc || !dashboardUrl}
+      />
+    );
+  }, [t, props.chartKind, updateKey, chartSrc, dashboardUrl]);
+
+  const popoutButton = React.useMemo(() => {
+    return (
+      <Button
+        key={1}
+        aria-label={t('CHART_CARD.BUTTONS.POPOUT.LABEL', { chartKind: props.chartKind })}
+        onClick={popout}
+        variant="plain"
+        icon={<ExternalLinkAltIcon />}
+        isDisabled={!chartSrc || !dashboardUrl}
+      />
+    );
+  }, [t, props.chartKind, popout, chartSrc, dashboardUrl]);
+
+  const actions = React.useMemo(() => {
+    const a = props.actions || [];
+    return [resyncButton, popoutButton, ...a];
+  }, [props.actions, resyncButton, popoutButton]);
+
+  const header = React.useMemo(() => {
+    const isWide = props.span > 4;
+    const style = {
+      marginBottom: isWide ? '-2em' : '',
+    };
+    if (hasRecording) {
+      return (
+        <>
+          <CardHeader style={style}>
+            <CardActions>{actions}</CardActions>
+          </CardHeader>
+        </>
+      );
+    } else {
+      return (
+        <>
+          <CardHeader style={style}>
+            <CardTitle>{props.chartKind}</CardTitle>
+            <CardActions>{props.actions}</CardActions>
+          </CardHeader>
+        </>
+      );
+    }
+  }, [props.actions, props.span, props.chartKind, hasRecording, actions]);
+
+  const handleCreateRecording = React.useCallback(() => {
+    history.push({
+      pathname: '/recordings/create',
+      state: {
+        restartExisting: true,
+        name: RECORDING_NAME,
+        templateName: 'Profiling',
+        templateType: 'TARGET',
+        labels: [{ key: 'origin', value: RECORDING_NAME }],
+        duration: -1,
+        // TODO these are arbitrary defaults that will be set in the recording creation form.
+        // Should these values be inferred in some more intelligent way?
+        maxAge: 120, // seconds
+        maxSize: 100 * 1024 * 1024, // bytes
+      } as CreateRecordingProps,
+    });
+  }, [history]);
+
+  return (
+    <DashboardCard
+      id={props.chartKind + '-chart-card'}
+      dashboardId={props.dashboardId}
+      cardSizes={ChartCardSizes}
+      isCompact
+      style={cardStyle}
+      cardHeader={header}
+    >
+      <CardBody>
+        {hasRecording ? (
+          <iframe key={key} style={{ height: '100%', width: '100%' }} src={chartSrc} />
```

Comment:
Maybe have a `LoadingView` component render during the card chart init when a state `isLoading` is true? I think this is possible with the `onLoad` prop on the `iframe` component, the only problem is that I believe the onload requires the iframe to be rendered first so we cant do `isLoading ? <LoadingView /> : <iframe ... >`. Maybe hiding the iframe with display none may be better until it is finished loading.

Reply:
The `onLoad` of an `iframe` fires when the source document has loaded, but since the Grafana chart component inside the frame has to also execute a bunch of JS and load data async from the datasource, I think the `<LoadingView>` would only appear for a brief flash until the embedded HTML has been loaded by the browser. I can give it a try and see how it feels in practice - I guess for slower connections between the browser and the Grafana dashboard in the cluster, this might still be visible for a while.

Reply:
The latest changes have a `<LoadingView>` but it triggers and is removed separately from the iframe loading cycle. It now actually appears when the controller is in the middle of determining whether the selected target has a suitable source recording. If the controller determines that it does then the card component should just about immediately mount the iframe and remove the loading spinner, so there may possibly be a lag time in between the loading spinner disappearing and any content actually rendering within the iframe. But this solution was also nicely unit testable, so I'm going with it for now. I can piece something together later to bridge that gap longer if we think it's needed and I figure out that onload event for tests.

Reply:
Oopps I miss this comment. Make sense to me to keep the loading state.